### ISSUE TITLE
DirChanged should fire whenever the result of getcwd() changes

### ISSUE THREAD

**Issue by justinmk** [neovim/neovim#6054]
`:help DirChanged` says:

> After the current-directory was changed.

But DirChanged only fires on an explicit call to e.g. `:lcd`. It should _also_ fire when user focuses a window or tab which causes CWD to change. Example:

```
$ strace nvim -u NONE +'split|lcd ..|wincmd w' +qa 2>&1|grep chdir
chdir("/home/vagrant/neovim")           = 0
chdir("/home/vagrant")                  = 0
chdir("/home/vagrant/neovim")           = 0
chdir("..")                             = 0
chdir("/home/vagrant/neovim")           = 0  # DirChanged should fire here. 
```

see https://github.com/alexgenco/neovim-ruby/pull/22#issuecomment-277394957

cc @mhinz

**Comment by alexgenco** [comment]
I feel like the `scope` argument might make less sense in the case where it's not triggered by an explicit `cd`, `lcd`, or `tcd`. I suppose in your example it would be `win`, but I could imagine cases where it's not as clear cut, like switching to a split whose pwd was determined by a global directory change.

**Comment by justinmk** [comment]
Perhaps it's reasonable that "scope" is simply "the current effective scope" in those cases.

**Comment by justinmk** [comment]
> like switching to a split whose pwd was determined by a global directory change.

In that case the scope would be "global", because that is the effective scope. Whether anyone will ever find this useful, I'm not sure. But the internal logic is such that exposing the conclusion is straightforward, so that's what we do.

This is implemented in #6262.
